barretenberg/cpp/src/barretenberg/vm/tests/avm_kernel.test.cpp

```diff
@@ -662,6 +698,32 @@ TEST_F(AvmKernelNegativeTests, incorrectIaStorageAddress)
     negative_test_incorrect_ia_kernel_lookup(apply_opcodes, checks, incorrect_ia, "PERM_MAIN_MEM_A");
 }
 
+TEST_F(AvmKernelNegativeTests, incorrectIaFunctionSelector)
+{
+    uint32_t dst_offset = 42;
+    FF incorrect_ia = FF(69);
+
+    // We test that the sender opcode is inlcuded at index x in the public inputs
+    auto apply_opcodes = [=](AvmTraceBuilder& trace_builder) {
+        trace_builder.op_function_selector(/*indirect*/ false, dst_offset);
+    };
+    auto checks = [=](bool indirect, const std::vector<Row>& trace) {
+        auto row = std::ranges::find_if(
+            trace.begin(), trace.end(), [](Row r) { return r.main_sel_op_function_selector == FF(1); });
+        EXPECT_TRUE(row != trace.end());
+
+        expect_row(
+            row,
+            /*kernel_in_offset=*/FUNCTION_SELECTOR_SELECTOR,
+            /*ia=*/incorrect_ia, // Note the value generated above for public inputs is the same as the index read + 1
+            /*ind_a*/ indirect,
+            /*mem_addr_a=*/dst_offset,
+            /*w_in_tag=*/AvmMemoryTag::U32);
+    };
+
+    negative_test_incorrect_ia_kernel_lookup(apply_opcodes, checks, incorrect_ia, "PERM_MAIN_MEM_A");
```

Comment:
We are not testing the right violation. This is testing that the memory trace is not compliant with the main trace memory operations pertaining to intermediate register IA. The whole list of such negative tests need to be adjusted as follows:

- Mutate the memory trace with the "incorrect value" (Then PERM_MAIN_MEM_A should pass.)
- Adjust the expected lookup violation.

@dbanks12 I think this is a good learning exercise and I can assist you with that.